lib/collection/src/grouping/aggregator.rs

```diff
@@ -0,0 +1,332 @@
+use std::collections::{HashMap, HashSet};
+
+use segment::types::{ExtendedPointId, ScoredPoint};
+
+type Hits = HashSet<ScoredPoint>;
+
+#[allow(dead_code)] // temporary
+struct GroupsAggregator {
+    groups: HashMap<String, Hits>,
+    max_group_size: usize,
+    grouped_by: String,
+    max_groups: usize,
+}
+
+#[allow(dead_code)] // temporary
+impl GroupsAggregator {
+    fn new(groups: usize, group_size: usize, grouped_by: String) -> Self {
+        Self {
+            groups: HashMap::with_capacity(groups),
+            max_group_size: group_size,
+            grouped_by,
+            max_groups: groups,
+        }
+    }
+
+    /// Adds a point to the group that corresponds based on the group_by field, assumes that the point has the group_by field
+    fn add_point(&mut self, point: &ScoredPoint) {
+        // if the key contains multiple values, grabs the first one
+        let group_value = point
+            .payload
+            .as_ref()
+            .unwrap()
+            .get_value(&self.grouped_by)
+            .values()
+            .first()
+            .unwrap()
+            .to_string()
+            .trim_matches('"')
+            .to_owned();
+
+        println!("point_id: {:?}, group_value: {:#?}", point.id, group_value);
+
+        if !self.groups.contains_key(&group_value) && self.groups.len() >= self.max_groups {
+            return;
+        }
+
+        let group = self
+            .groups
+            .entry(group_value)
+            .or_insert_with(|| HashSet::with_capacity(self.max_group_size));
+
+        if group.len() >= self.max_group_size {
+            return;
+        }
+
+        group.insert(point.clone());
+    }
+
+    /// Adds multiple points to the group that they corresponds based on the group_by field, assumes that the points always have the grouped_by field
+    fn add_points(&mut self, points: &[ScoredPoint]) {
+        print_points(points);
+        points.iter().for_each(|point| self.add_point(point));
+        for group in self.groups.iter() {
+            println!("group: {:#?}", group);
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.groups.len()
+    }
+
+    // gets the keys of the groups that have less than the max group size
+    fn keys_of_unfilled_groups(&self) -> Vec<String> {
+        self.groups
+            .iter()
+            .filter(|(_, hits)| hits.len() < self.max_group_size)
+            .map(|(key, _)| key.clone())
+            .collect()
+    }
+
+    // gets the keys of the groups that have reached or exceeded the max group size
+    fn keys_of_filled_groups(&self) -> Vec<String> {
+        self.groups
+            .iter()
+            .filter(|(_, hits)| hits.len() >= self.max_group_size)
```

Comment:
any reason to keep in memory more than `max.group_size` hits per group?

Reply:
Maybe the `>=` operator is confusing, I'll replace it for a `==`, we shouldn't have groups that exceed this